### PR TITLE
Add Material Support to Globe

### PR DESCRIPTION
**Pull request by jasonbeverage**
Hi all!

This PR adds Material support to the Globe, as well as adding new height and slope values to the czm_materialInput struct so that it can be used as input when generating the materials.

I've added a new sandcastle example that you can play with [here](https://cesiumglobematerial.netlify.com/apps/sandcastle/?src=Globe%20Materials.html&label=Showcases)

The main goal of this effort was to visualize changes in slope on the terrain, but I tried to make it more generic by utilizing the existing Material support so we could do some more interesting effects in the future.  I've added some new materials that you can see in use in the sandcastle example
* ElevationContour - Displays elevation contour lines on the terrain with the spacing controled via a uniform.  You can drag the slider in the sandcastle example to change the spacing.
* ElevationRamp - Colors the terrain by applying a color ramp to the elevation.  The elevation is scaled between minHeight and maxHeight uniforms and then does a texture lookup into the ramp to get the color.  This ended up being really cool because you can generate nice color ramps programatically using a canvas and the createLinearGradient function.
* SlopeRamp - Colors the terrain by applying a color ramp to the slope.

Some notes on the implementation:
*  I tried to keep the changes as minimal as possible and just re-use the great work already done in the existing Material framework.
*  The Globe now has a new material property which you can set to any Material you like.
*  Globe has a new dirtyShaders function that is called when you change the material so the correct material shader is injected in the globe shader.
*  If there is no material there shouldn't be a performance penalty b/c it's wrapped in a APPLY_MATERIAL define which is only set if there is material present.
*  GlobeSurfaceShader has a new material property so it can detect when the material has changed and a new shader needs to be generated.
*  GlobeSurfaceTileProvider has a new uniformMap property that is set to the materials uniforms so they are incorporated in the tile draw calls.
*  Currently the material only effects the diffuse color of the terrain.  The original imagery color is blended with the materials color based on the materials alpha.  This allows you to return an alpha value less than 1 to get the original imagery color to blend though.  For example, try changing one of the color stops in the color ramp to "transparent" or "rgba(255,0,0,0.5)" and see the imagery come through.

Thanks!

Jason

heightramp

contourmap

### CODE REVIEW

**Comment by pjcozzi** [comment]
@jasonbeverage this is super exciting, thanks again!

Here's Seneca Rocks where I use to rock climb:

image

For the contours, consider adding an optional where they "maintain" constant pixel width using screen-space partial derivatives.  Perhaps something like this: https://github.com/AnalyticalGraphicsInc/cesium/blob/a81f2e7de64c2348558046b5e8e685278fde0dcf/Source/Shaders/Materials/GridMaterial.glsl#L23

@lilleyse can you take a first pass at this?  I would like to do a quick review before merging.

**Comment by jasonbeverage** [comment]
That's awesome you used to rock climb in Seneca Rocks!  I'm from West
Virginia so I've been there a bunch as a kid and I went camping and trout
fishing in that area last year with my uncle.

File: Source/Scene/GlobeSurfaceTileProvider.js [hunk 1]
```diff
@@ -1262,6 +1264,10 @@ define([
             uniformMapProperties.minMaxHeight.y = encoding.maximumHeight;
             Matrix4.clone(encoding.matrix, uniformMapProperties.scaleAndBias);
 
+            if (tileProvider.uniformMap) {
+                 uniformMap = combine(uniformMap, tileProvider.uniformMap);
+            }
```
Comment (lilleyse) [inline Source/Scene/GlobeSurfaceTileProvider.js]: It would be best to combine uniforms elsewhere since this will allocate a uniform map object every frame.

I wonder if this could be moved to the area near `if (tileProvider._drawCommands.length <= tileProvider._usedDrawCommands) {`. Can it detect when the material is dirty and combine up there?

File: Source/Shaders/GlobeVS.glsl [hunk 2]
```diff
@@ -21,6 +21,8 @@ varying vec3 v_positionEC;
 varying vec3 v_textureCoordinates;
 varying vec3 v_normalMC;
 varying vec3 v_normalEC;
+varying float v_slope;
```
Comment (lilleyse) [inline Source/Shaders/GlobeVS.glsl]: Everything in here should also be surrounded by the `APPLY_MATERIAL` define too.

File: Source/Shaders/GlobeFS.glsl [hunk 3]
```diff
@@ -61,6 +61,8 @@ varying vec3 v_positionEC;
 varying vec3 v_textureCoordinates;
 varying vec3 v_normalMC;
 varying vec3 v_normalEC;
+varying float v_height;
+varying float v_slope;
```
Comment (lilleyse) [inline Source/Shaders/GlobeFS.glsl]: Surround these with the `APPLY_MATERIAL` define.

File: Source/Scene/Globe.js [hunk 4]
```diff
@@ -551,6 +570,35 @@ define([
...
+        });
+
+        var fragmentSources = [];
+        var fragmentDefines = [];
+        if (this._material) {
+            fragmentSources.push(this._material.shaderSource);
+            fragmentDefines.push('APPLY_MATERIAL');
+
+            // Set the material uniform map to the materials
+            this._surface._tileProvider.uniformMap = this._material._uniforms;
```
[the remark this answers is not in the document]
Comment (lilleyse) [reply Source/Scene/Globe.js]: Also a better name would be `materialUniformMap`.

File: Source/Scene/GlobeSurfaceTileProvider.js [hunk 5]
```diff
@@ -1262,6 +1264,10 @@ define([
             uniformMapProperties.minMaxHeight.y = encoding.maximumHeight;
             Matrix4.clone(encoding.matrix, uniformMapProperties.scaleAndBias);
 
+            if (tileProvider.uniformMap) {
```
Comment (lilleyse) [inline Source/Scene/GlobeSurfaceTileProvider.js]: Use `defined`.

**Comment by lilleyse** [comment]
Added some style tweaks in https://github.com/AnalyticalGraphicsInc/cesium/pull/5919/commits/60c0121ed2b257e6bd5ff7dadffeb0ff4335ff7c

**Comment by lilleyse** [comment]
Remaining TODO:

* [ ] Tests
* [ ] Improve contour lines: https://github.com/AnalyticalGraphicsInc/cesium/pull/5919/#issuecomment-341295918

File: Apps/Sandcastle/gallery/Globe Materials.html [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer');
+var cesiumTerrainProviderMeshes = new Cesium.CesiumTerrainProvider({
+    url : 'https://assets.agi.com/stk-terrain/v1/tilesets/world/tiles',
+    requestWaterMask : true,
+    requestVertexNormals : true
+});
+viewer.terrainProvider = cesiumTerrainProviderMeshes;
+
+viewer.scene.globe.enableLighting = true;
```
[the remark this answers is not in the document]
Comment (hpinkos) [reply Apps/Sandcastle/gallery/Globe Materials.html]: The color ramps are better with lighting so you can see the terrain features.  The terrain all blends together without the shadows.

File: Source/Scene/Material.js [hunk 7]
```diff
@@ -1470,5 +1492,59 @@ define([
...
+     * @readonly
+     */
+    Material.ElevationRampType = 'ElevationRamp';
+    Material._materialCache.addMaterial(Material.ElevationRampType, {
+        fabric : {
+            type : Material.ElevationRampType,
+            uniforms : {
+                image: Material.DefaultImageId,
+                minHeight: 0.0,
+                maxHeight: 10000.0,
```
Comment (hpinkos) [inline Source/Scene/Material.js]: Change these to `minimumHeight` and `maximumHeight`.  We try to avoid abbreviations so there's no confusion

File: Apps/Sandcastle/gallery/Globe Materials.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (lilleyse) [reply Apps/Sandcastle/gallery/Globe Materials.html]: The lighting does look fine most of the time, it's mainly when its completely dark that it's hard to see what's going on.

[the remark this answers is not in the document]
Comment (hpinkos) [reply Apps/Sandcastle/gallery/Globe Materials.html]: I disagree.  For example, there are some random Appalachian mountains with and without lighting:

image

image

Having the lighting enabled makes a big difference.
To make sure the lighting always looks good with the preset views, we can add a line to set the clock time:
`viewer.clockViewModel.currentTime = Cesium.JulianDate.fromDate(new Date(2017, 8, 22, 0, 0, 0));`

[the remark this answers is not in the document]
Comment (lilleyse) [reply Apps/Sandcastle/gallery/Globe Materials.html]: Ok yeah, setting the clock is good. I should have been clearer, my main issue was that whenever I would open the demo it would be night at Everest and just hard to see anything.

File: Source/Scene/Material.js (hunk 7 above)
Comment (hpinkos) [reply Source/Scene/Material.js]: Remove the dangling comma here

**Comment by pjcozzi** [comment]
@jasonbeverage are you able to make the suggested changes to get this over the finish line for the upcoming 1.40 release? 😄

**Comment by jasonbeverage** [comment]
Hey Patrick I'll work on getting these addressed real soon.  Thanks!

**Comment by pjcozzi** [comment]
Very nice, rock on @jasonbeverage!

File: Apps/Sandcastle/gallery/Globe Materials.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (jasonbeverage) [reply Apps/Sandcastle/gallery/Globe Materials.html]: I've pushed a change to set the clock time.

File: Source/Scene/Globe.js [hunk 8]
```diff
@@ -525,6 +540,10 @@ define([
             return;
         }
 
+        if (this._material) {
```
[the remark this answers is not in the document]
Comment (jasonbeverage) [reply Source/Scene/Globe.js]: Fixed.

File: Source/Scene/Globe.js [hunk 9]
```diff
@@ -551,6 +570,35 @@ define([
...
+     * @private
+     */
+    Globe.prototype.dirtyShaders = function() {
+        this._surfaceShaderSet.baseVertexShaderSource = new ShaderSource({
+            sources : [GroundAtmosphere, GlobeVS]
+        });
+
+        var fragmentSources = [];
+        var fragmentDefines = [];
+        if (this._material) {
```
[the remark this answers is not in the document]
Comment (jasonbeverage) [reply Source/Scene/Globe.js]: Fixed.

File: Source/Scene/GlobeSurfaceTileProvider.js (hunk 5 above)
Comment (jasonbeverage) [reply Source/Scene/GlobeSurfaceTileProvider.js]: Fixed.

File: Source/Shaders/GlobeVS.glsl (hunk 2 above)
Comment (jasonbeverage) [reply Source/Shaders/GlobeVS.glsl]: Fixed

File: Source/Shaders/GlobeFS.glsl (hunk 3 above)
Comment (jasonbeverage) [reply Source/Shaders/GlobeFS.glsl]: Fixed

File: Source/Scene/Material.js (hunk 7 above)
Comment (jasonbeverage) [reply Source/Scene/Material.js]: Fixed.

**Comment by jasonbeverage** [comment]
I just pushed a change to the ElevationContourMaterial that uses the code from @pasu to implement the screen space partial derivative based contouring.  I also added a new lineThickness uniform so you can control the thickness of the line.  It looks pretty good to me.

I think my recent changes addresses everything mentioned here.  Let me know if you see anything else!

**Comment by jasonbeverage** [comment]
New sandcastle demo looks sweet @hpinkos !

**Comment by hpinkos** [comment]
Thanks @jasonbeverage!  I just added a slider for the line thickness too =)

File: CHANGES.md [hunk 10]
```diff
@@ -3,6 +3,7 @@ Change Log
 
 ### 1.40 - 2017-12-01
 
+* Adds Material support to Globe. [#5919](https://github.com/AnalyticalGraphicsInc/cesium/pull/5919)
```
Comment (pjcozzi) [inline CHANGES.md]: Can this link to the future Sandcastle example that will live here: https://cesiumjs.org/Cesium/Apps/Sandcastle/

File: Apps/Sandcastle/gallery/Globe Materials.html [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+
+var elevationRamp = [0.0, 0.045, 0.1, 0.15, 0.37, 0.54, 1.0];
+var slopeRamp = [0.0, 0.29, 0.5, Math.sqrt(2)/2, 0.87, 0.91, 1.0];
+function getColorRamp(selectedShading) {
+    var ramp = document.createElement('canvas');
+    ramp.width = 100;
+    ramp.height = 1;
+    var ctx = ramp.getContext('2d');
+
+    var values = selectedShading === 'elevation' ? elevationRamp : slopeRamp;
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Globe Materials.html]: @hpinkos can the color ramp be tweaked to be a bit less subtle for the default view?

image

Comment (pjcozzi) [reply Apps/Sandcastle/gallery/Globe Materials.html]: Or change the default view.

File: Source/Scene/Globe.js [hunk 12]
```diff
@@ -550,6 +569,36 @@ define([
         }
     };
 
+    /**
+     * @private
+     */
+    Globe.prototype.dirtyShaders = function() {
```
Comment (pjcozzi) [inline Source/Scene/Globe.js]: We would generally name this a verb like `makeShadersDirty` and not make it on the `Globe` object; instead, it would be a stand-alone function at file level scope and then we pass in a `Globe` object.  This provides better encapsulation.

File: Source/Scene/Globe.js [hunk 13]
```diff
@@ -550,6 +569,36 @@ define([
...
+        var defines = [];
+
+        var fragmentSources = [];
+        if (defined(this._material)) {
+            fragmentSources.push(this._material.shaderSource);
+            defines.push('APPLY_MATERIAL');
+
+            // Set the material uniform map to the materials
```
Comment (pjcozzi) [inline Source/Scene/Globe.js]: Probably don't need this comment.

File: Source/Scene/Globe.js [hunk 14]
```diff
@@ -550,6 +569,36 @@ define([
...
+            this._surface._tileProvider.uniformMap = this._material._uniforms;
+        } else {
+            this._surface._tileProvider.uniformMap = undefined;
+        }
+        fragmentSources.push(GlobeFS);
+
+        this._surfaceShaderSet.baseVertexShaderSource = new ShaderSource({
+            sources : [GroundAtmosphere, GlobeVS],
+            defines: defines
```
Comment (pjcozzi) [inline Source/Scene/Globe.js]: Throughout, we would generally write `defines :` with a space before the semicolon.  The auto formatter in your IDE probably does this.  WebStorm does.

File: Source/Scene/Material.js [hunk 15]
```diff
@@ -226,6 +232,23 @@ define([
      *      <li><code>outlineColor</code>: diffuse color and alpha for the outline.</li>
      *      <li><code>outlineWidth</code>: width of the outline in pixels.</li>
      *  </ul>
+     *  <li>ElevationContour</li>
+     *  <ul>
+     *      <li><code>color</code>: color and alpha for the contour line.</li>
+     *      <li><code>spacing</code>: spacing for contour lines in meters.</li>
+     *      <li><code>lineThickness</code>: Number specifying the thickness of the grid lines in pixels.</li>
```
Comment (pjcozzi) [inline Source/Scene/Material.js]: To be consistent with the rest of the Cesium API, can this be "width" instead of "thickness?"

File: Source/Shaders/Builtin/Structs/materialInput.glsl [hunk 16]
```diff
@@ -10,6 +10,8 @@
  * @property {vec3} normalEC Unperturbed surface normal in eye coordinates.
  * @property {mat3} tangentToEyeMatrix Matrix for converting a tangent space normal to eye space.
  * @property {vec3} positionToEyeEC Vector from the fragment to the eye in eye coordinates.  The magnitude is the distance in meters from the fragment to the eye.
+ * @property {float} height The height of the terrain.  Only available for globe materials.
```
Comment (pjcozzi) [inline Source/Shaders/Builtin/Structs/materialInput.glsl]: Can this be more explicit?  I assume the height in meters above or below the WGS84 ellipsoid?

File: Source/Shaders/Builtin/Structs/materialInput.glsl [hunk 17]
```diff
@@ -10,6 +10,8 @@
...
+ * @property {float} slope The slope of the terrain normalized from 0 to 1.  Only available for globe materials.
```
Comment (pjcozzi) [inline Source/Shaders/Builtin/Structs/materialInput.glsl]: Can this be more explicit, e.g., 0 is flat and 1 is vertical or vice versa?

File: Source/Shaders/GlobeVS.glsl [hunk 18]
```diff
@@ -167,4 +172,11 @@ void main()
     v_rayleighColor = atmosColor.rayleigh;
     v_distance = length((czm_modelView3D * vec4(position3DWC, 1.0)).xyz);
 #endif
+
+#ifdef APPLY_MATERIAL
+    vec3 finalNormal = normalize(v_normalMC);
```
Comment (pjcozzi) [inline Source/Shaders/GlobeVS.glsl]: No need to normalize, it is already done: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Shaders/Builtin/Functions/octDecode.glsl#L24

File: Source/Shaders/GlobeVS.glsl [hunk 19]
```diff
@@ -167,4 +172,11 @@ void main()
...
+    vec3 worldNormal = normalize(v_positionMC.xyz);
```
Comment (pjcozzi) [inline Source/Shaders/GlobeVS.glsl]: Here and the line above, we generally try to treat varyings in a vertex shader as "write once, don't read" - more for style - I don't think it matters for performance.

Here, the return from `czm_octDecode` could be assigned to a local, and `position3DWC` can be used instead of `v_positionMC`.

Comment (pjcozzi) [reply Source/Shaders/GlobeVS.glsl]: Also `worldNormal` is a confusing name - since it is not the terrain normal in world coordinate - perhaps rename to `ellipsoidNormal` since this is normal to the ellipsoid?

**Comment by jasonbeverage** [comment]
Hey @pjcozzi I'll try to make those changes tonight or tomorrow.  Do you have an example of something I could use for inspiration for the unit tests?  I'm still not the best at writing them :)

I'm heading out of town for the IITSEC conference on Sunday and won't be back until the end of the week, so I'll try to get what I can wrapped up before I head out.

**Comment by pjcozzi** [comment]
@jasonbeverage awesome - it is deeply appreciated!

I'm not sure that it is the _best_ example, but here is a simple example for rendering a globe in a unit test:

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Scene/SceneSpec.js#L488-L505

For your unit tests, just cover rendering with each type of material and all the edge cases, e.g., changing material, going from no material to material, from material to no material, perhaps changing uniforms, etc.

The test guiding is pretty is good:

https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide#testing-guide

Also:
* Please keep your eye out for cool Cesium apps at IITSEC - would love to hear what you find!
* What is your next Cesium feature? 😄

File: Apps/Sandcastle/gallery/Globe Materials.html (hunk 11 above)
Comment (jasonbeverage) [reply Apps/Sandcastle/gallery/Globe Materials.html]: I changed the default view to be of the Himalayas

**Comment by jasonbeverage** [comment]
I think I've addressed all of your comments except for the unit tests.

**Comment by jasonbeverage** [comment]
I just added a few unit tests that appear to pass :)

I'm heading out for the rest of the day and will be gone until next Wednesday but I'll be answering email and checking this PR if you need anything ;)

**Comment by pjcozzi** [comment]
Very nice, thanks again @jasonbeverage.  Have a great conference!

**Comment by jasonbeverage** [comment]
Thanks for merging this!  I'll hopefully have at least one more little PR before the end of the year :)

**Comment by MapGiser** [comment]
> Hi, I also did this job but I just modified the globefs directly,  I never thought adding material to support globe, it is a decent design, I gained it.
> Here is my little suggestion to set the contour line.
> Source/Shaders/Materials/ElevationContourMaterial.glsl:
> czm_material material = czm_getDefaultMaterial(materialInput);
> float dContourInterval = 100.0;// the interval value of contour line
> float distanceToContour = mod(materialInput.height, dContourInterval);
> float dxc = abs(dFdx(materialInput.height));
> float dyc = abs(dFdy(materialInput.height));
> float dF = max(dxc, dyc);
> material.alpha = (distanceToContour < dF) ? 1.0 : 0.0;
> material.diffuse = color.rgb;
> return material;
I want to draw counterline in restrict area, can you give me some advise?thanks

**Comment by Abhishek950650** [comment]
@jasonbeverage. Hello i am trying to implement Cesium globe materials slope, elevation, aspect features in my local project but i am facing issue in my project if i apply any these options my globe is not showing perfect output its look like  only single color on my whole globe 🌐 like red or blue or green
How i resolve this issue.. 

Thanks

**Comment by OmarShehata** [comment]
@Abhishek950650 I would ask your question in the Cesium community forum (https://community.cesium.com/) so others can take a look and help out. Include more information like a Sandcastle and how you're using these materials (see https://community.cesium.com/t/how-to-share-custom-sandcastle-examples/9828)
